File: link_climate.py
#!/usr/bin/env python3
"""
Link biological / point observations to climate exposure windows.

Take a CSV where **each row has its own location and its own date** (a biological
sample, a clinical visit, a survey point) and write out the same rows enriched
with climate aggregates over the days leading up to each observation.

Examples
--------
Basic (temperature, default windows of 1, 7 and 30 days):

    python link_climate.py --input samples.csv --output samples_climate.csv \
        --lat-col latitude --lon-col longitude --date-col date

Multiple variables and custom windows, plus a 3-day lag (exclude the 3 days
immediately before the event):

    python link_climate.py -i samples.csv -o out.csv \
        --variables temperature,precipitation,humidity \
        --windows 7,14,30,60 --lag 3

The input CSV needs latitude, longitude and date columns (names configurable).
Everything else in the file is carried through untouched.
"""

import argparse
import sys

import pandas as pd

# Allow running from anywhere: make the repo root importable.
import os
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from src.bio_climate_link import link_climate_dataframe, VARIABLE_COLUMNS  # noqa: E402


def _parse_int_list(text: str) -> list:
    return [int(x) for x in str(text).replace(" ", "").split(",") if x != ""]


def _parse_str_list(text: str) -> list:
    return [x for x in str(text).replace(" ", "").split(",") if x != ""]


def main():
    parser = argparse.ArgumentParser(
        description="Link biological observations (location + timestamp) to climate exposure windows.",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    parser.add_argument("-i", "--input", required=True, help="Input CSV path")
    parser.add_argument("-o", "--output", required=True, help="Output CSV path")
    parser.add_argument("--lat-col", default="latitude", help="Latitude column name")
    parser.add_argument("--lon-col", default="longitude", help="Longitude column name")
    parser.add_argument("--date-col", default="date", help="Observation date column (YYYY-MM-DD)")
    parser.add_argument("--id-col", default=None, help="Optional identifier column (carried through)")
    parser.add_argument(
        "--variables", default="temperature",
        help=f"Comma-separated climate variables. Options: {','.join(sorted(VARIABLE_COLUMNS))}",
    )
    parser.add_argument(
        "--windows", default="1,7,30",
        help="Comma-separated window lengths in days (each = N days ending on the observation date)",
    )
    parser.add_argument(
        "--lag", type=int, default=0,
        help="Days to shift each window back before the observation date",
    )
    parser.add_argument(
        "--buffer-km", type=float, default=0.0,
        help="Spatial buffer radius in km (0 = exact point)",
    )
    parser.add_argument(
        "--coord-decimals", type=int, default=3,
        help="Round coordinates to this many decimals when grouping sites (3 ~ 100 m)",
    )
    parser.add_argument("--project-id", default=None, help="Google Earth Engine / Cloud project id")
    parser.add_argument("--no-same-day", action="store_true", help="Do not emit same-day raw values")
    parser.add_argument("--no-coverage", action="store_true", help="Do not emit data-coverage columns")

    args = parser.parse_args()

    try:
        df = pd.read_csv(args.input)
    except Exception as exc:  # noqa: BLE001
        print(f"❌ Could not read input CSV '{args.input}': {exc}")
        sys.exit(1)

    print(f"📄 Read {len(df)} rows from {args.input}")
    print("🔗 Linking climate exposure windows — this queries Google Earth Engine per site...")

    try:
        enriched = link_climate_dataframe(
            df,
            lat_col=args.lat_col,
            lon_col=args.lon_col,
            date_col=args.date_col,
            id_col=args.id_col,
            variables=_parse_str_list(args.variables),
            windows=_parse_int_list(args.windows),
            lag=args.lag,
            buffer_km=args.buffer_km,
            coord_decimals=args.coord_decimals,
            include_same_day=not args.no_same_day,
            add_coverage=not args.no_coverage,
            project_id=args.project_id,
            progress=True,
        )
    except (KeyError, ValueError, RuntimeError) as exc:
        print(f"❌ {exc}")
        sys.exit(1)

    enriched.to_csv(args.output, index=False)
    new_cols = [c for c in enriched.columns if c not in df.columns]
    print(f"✅ Wrote {len(enriched)} rows and {len(new_cols)} new climate columns to {args.output}")
    print(f"   New columns: {', '.join(new_cols)}")


if __name__ == "__main__":
    main()

File: r/climatelink/R/link_climate.R
#' Link biological observations to climate exposure windows
#'
#' For each row of `data` (which must carry a latitude, a longitude and a date),
#' attach climate aggregates over the days leading up to the observation. Earth
#' Engine is queried once per unique site, then windows are computed locally, so
#' repeated field sites are cheap.
#'
#' Output columns are named `{variable}_{aggregation}_{N}d` — for example
#' `tmean_celsius_mean_30d` is the mean daily mean-temperature over the 30 days
#' ending on the observation date. Same-day raw values are named `{variable}`
#' (e.g. `tmean_celsius`). A `coverage_{N}d` column reports the fraction of each
#' window's days that had data (watch for values < 1, which flag gaps or masked
#' pixels).
#'
#' @param data A data.frame of observations.
#' @param lat_col,lon_col Names of the latitude / longitude columns (decimal
#'   degrees, WGS84).
#' @param date_col Name of the observation-date column. Dates, POSIXct or
#'   "YYYY-MM-DD" character are all accepted.
#' @param id_col Optional identifier column name (carried through untouched).
#' @param variables Character vector of climate variables to link. One or more of
#'   `"temperature"`, `"precipitation"`, `"humidity"`, `"wind"`, `"solar"`,
#'   `"pressure"`, `"evapotranspiration"`.
#' @param windows Integer vector of window lengths in days. A window of `N` covers
#'   the `N` days ending on (and including) the observation date.
#' @param lag Days to shift every window back before the observation date
#'   (e.g. `lag = 1` ends each window the day before the event).
#' @param buffer_km Spatial buffer radius in km. `0` samples the exact pixel;
#'   a positive value averages over land pixels within the buffer, which is the
#'   robust choice for coastal or water-adjacent sites where the exact pixel may
#'   be masked (returns `NA`).
#' @param aggregations Optional named list overriding the default aggregation(s)
#'   per raw column, e.g. `list(tmean_celsius = c("mean", "max"))`.
#' @param include_same_day If `TRUE`, also return the raw same-day value of each
#'   variable.
#' @param add_coverage If `TRUE`, add data-coverage columns.
#' @param coord_decimals Decimals to round coordinates to when grouping sites
#'   (3 is roughly 100 m).
#' @param project_id Google Earth Engine / Cloud project id to initialise with.
#' @param progress If `TRUE`, print per-site progress.
#'
#' @return A data.frame: `data` with the climate columns appended (original row
#'   order preserved).
#'
#' @examples
#' \dontrun{
#' obs <- data.frame(
#'   id = c("a", "b"),
#'   latitude = c(-26.2678, -33.9249),
#'   longitude = c(27.8607, 18.4241),
#'   date = as.Date(c("2021-01-15", "2021-03-10"))
#' )
#' linked <- link_climate(
#'   obs,
#'   variables = c("temperature", "precipitation"),
#'   windows = c(1, 7, 30),
#'   buffer_km = 10,          # robust for the coastal Cape Town point
#'   project_id = "your-gee-project"
#' )
#' }
#' @export
link_climate <- function(data,
                         lat_col = "latitude",
                         lon_col = "longitude",
                         date_col = "date",
                         id_col = NULL,
                         variables = "temperature",
                         windows = c(1, 7, 30),
                         lag = 0,
                         buffer_km = 0,
                         aggregations = NULL,
                         include_same_day = TRUE,
                         add_coverage = TRUE,
                         coord_decimals = 3,
                         project_id = NULL,
                         progress = TRUE) {
  if (!is.data.frame(data)) stop("'data' must be a data.frame", call. = FALSE)
  for (col in c(lat_col, lon_col, date_col)) {
    if (!col %in% names(data)) {
      stop("Column '", col, "' not found in 'data'. Available: ",
           paste(names(data), collapse = ", "), call. = FALSE)
    }
  }

  # Normalise dates to "YYYY-MM-DD" character so reticulate ships them cleanly to
  # pandas (Date/POSIXct can otherwise arrive as numbers).
  data <- as.data.frame(data)
  if (inherits(data[[date_col]], c("Date", "POSIXct", "POSIXt"))) {
    data[[date_col]] <- format(as.Date(data[[date_col]]), "%Y-%m-%d")
  } else {
    data[[date_col]] <- as.character(data[[date_col]])
  }

  mod <- .get_module()

  py_aggs <- if (is.null(aggregations)) NULL else lapply(aggregations, as.list)

  result <- mod$link_climate_dataframe(
    df = reticulate::r_to_py(data),
    lat_col = lat_col,
    lon_col = lon_col,
    date_col = date_col,
    id_col = id_col,
    variables = as.list(as.character(variables)),
    windows = as.list(as.integer(windows)),
    lag = as.integer(lag),
    aggregations = py_aggs,
    include_same_day = isTRUE(include_same_day),
    add_coverage = isTRUE(add_coverage),
    buffer_km = as.numeric(buffer_km),
    coord_decimals = as.integer(coord_decimals),
    project_id = project_id,
    progress = isTRUE(progress)
  )

  out <- reticulate::py_to_r(result)
  rownames(out) <- NULL
  out
}

#' Link a CSV of observations to climate exposure windows
#'
#' File-in / file-out convenience wrapper around [link_climate()]. Reads
#' `input_path`, links it, and writes the enriched table to `output_path`.
#'
#' @param input_path,output_path Input and output CSV paths.
#' @param ... Passed to [link_climate()].
#' @return Invisibly, the enriched data.frame.
#' @export
link_climate_csv <- function(input_path, output_path, ...) {
  data <- utils::read.csv(input_path, stringsAsFactors = FALSE)
  linked <- link_climate(data, ...)
  utils::write.csv(linked, output_path, row.names = FALSE)
  message("Wrote ", nrow(linked), " rows to ", output_path)
  invisible(linked)
}
